src/app/file/gif_format.cpp

```diff
@@ -508,161 +512,136 @@ class GifDecoder {
     // We'll calculate the list of used colormap indexes in this
     // frameImage.
     PalettePicks usedEntries(ncolors);
-    if (isLocalColormap) {
-      // With this we avoid discarding the transparent index when a
-      // frame indicates that it uses a specific index as transparent
-      // but the image is completely opaque anyway.
-      if (!m_opaque && m_frameNum == 0 && m_localTransparentIndex >= 0 &&
-          m_localTransparentIndex < ncolors) {
-        usedEntries[m_localTransparentIndex] = true;
-      }
+    if (m_frameNum == 0 && !m_opaque &&
+        m_localTransparentIndex < ncolors &&
+        m_localTransparentIndex >= 0) {
+      m_bgIndex = m_localTransparentIndex;
+      m_sprite->setTransparentColor(m_bgIndex);
+      clear_image(m_currentImage.get(), m_bgIndex);
+      clear_image(m_previousImage.get(), m_bgIndex);
+      usedEntries[m_bgIndex] = true;
+    }
 
+    if (m_frameNum == 0 && !isLocalColormap)
+      // Mark all entries as used if the colormap is global.
+      usedEntries.all();
+    else {
       for (const auto& i : LockImageBits<IndexedTraits>(frameImage)) {
-        if (i >= 0 && i < ncolors && i != m_localTransparentIndex)
+        if (i >= 0 && i < ncolors)
           usedEntries[i] = true;
       }
     }
-    // Mark all entries as used if the colormap is global.
-    else {
-      usedEntries.all();
-    }
 
     // Number of colors (indexes) used in the frame image.
     int usedNColors = usedEntries.picks();
 
-    // Check if we need an extra color equal to the bg color in a
-    // transparent frameImage.
-    bool needsExtraBgColor = false;
-    bool needCheckLocalTransparent = m_bgIndex != m_localTransparentIndex ||
-                                     (ncolors > m_localTransparentIndex
-                                     && m_localTransparentIndex >= 0
-                                     && usedEntries[m_localTransparentIndex]);
-
-    if (m_sprite->pixelFormat() == IMAGE_INDEXED &&
-        !m_opaque &&
-        needCheckLocalTransparent) {
-      for (const auto& i : LockImageBits<IndexedTraits>(frameImage)) {
-        if (i == m_bgIndex) {
-          needsExtraBgColor = true;
-          break;
-        }
+    std::unique_ptr<Palette> palette;
+    if (m_frameNum == 0) {
+      int extraColor = m_localTransparentIndex >= ncolors ? 1 : 0;
+      palette.reset(new Palette(m_frameNum, usedNColors + extraColor));
+      if (!m_opaque && m_localTransparentIndex >= ncolors) {
+        // If m_localTransparentIndex is out of bounds, we'll define
+        // m_localTransparentIndex = last used color;
```

Comment:
Can the local transparent index be out of bounds (bounds of the local color map I guess?)? Why this happens? (buggy GIFs? maybe we can explain that some GIF files contains this kind of "out of bounds" transparent index).

Reply:
<img width="957" alt="Screenshot 2023-07-03 at 17 33 48" src="https://github.com/aseprite/aseprite/assets/32165475/6d22e97c-2a78-4480-8899-f19baf67c93d">